File: backend/app/schemas/scenario.py
"""
场景测试相关的 Pydantic 模型
"""
"""
版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。

本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。

授权商业应用请联系微信：huice666
"""


from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ==================== 场景模型 ====================
# fmt: off  MC80OmFIVnBZMlhsdktEbHVwNDZhRFpPUlE9PTo1NzJlNDBmNw==

class ScenarioBase(BaseModel):
    """场景基础模型"""
    name: str = Field(..., min_length=1, max_length=500, description="场景名称")
    description: Optional[str] = Field(None, description="场景描述")
    folder_id: Optional[UUID] = Field(None, description="所属文件夹 ID")
    global_variables: Dict[str, Any] = Field(default_factory=dict, description="全局变量")
    retry_count: int = Field(0, ge=0, description="重试次数")
    timeout_seconds: int = Field(300, ge=1, description="超时时间（秒）")
    parallel_execution: bool = Field(False, description="是否并行执行")


class ScenarioCreate(ScenarioBase):
    """创建场景请求"""
    pass


class ScenarioUpdate(BaseModel):
    """更新场景请求"""
    name: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = None
    global_variables: Optional[Dict[str, Any]] = None
    retry_count: Optional[int] = Field(None, ge=0)
    timeout_seconds: Optional[int] = Field(None, ge=1)
    status: Optional[str] = Field(None, pattern="^(draft|active|archived)$")


class ScenarioResponse(ScenarioBase):
    """场景响应"""
    id: UUID
    project_id: UUID
    identifier: str
    status: str
    total_steps: int
    last_run_status: Optional[str]
    last_run_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    created_by: Optional[UUID]
# pylint: disable  MS80OmFIVnBZMlhsdktEbHVwNDZhRFpPUlE9PTo1NzJlNDBmNw==

    class Config:
        from_attributes = True


class ScenarioListResponse(BaseModel):
    """场景列表响应"""
    total: int
    items: List[ScenarioResponse]


# ==================== 步骤模型 ====================

class DataMappingCreate(BaseModel):
    """创建数据映射请求"""
    source_type: str = Field(..., pattern="^(previous_response|variable|static)$")
    source_step_id: Optional[UUID] = None
    source_path: Optional[str] = None
    target_path: str = Field(..., description="目标路径，如 headers.Authorization")
    transform_expression: Optional[str] = None
    description: Optional[str] = None


class DataMappingResponse(BaseModel):
    """数据映射响应"""
    id: UUID
    source_type: str
    source_step_id: Optional[UUID]
    source_path: Optional[str]
    target_path: str
    transform_expression: Optional[str]
    description: Optional[str]

    class Config:
        from_attributes = True


class StepExtractor(BaseModel):
    """数据提取器"""
    name: str = Field(..., description="变量名")
    path: str = Field(..., description="JSONPath 路径")
    type: str = Field("jsonpath", description="提取器类型")


class StepAssertion(BaseModel):
    """断言"""
    type: str = Field(..., pattern="^(status|jsonpath|header)$")
    expected: Any
    path: Optional[str] = None
    operator: str = Field("eq", description="比较运算符")

# pylint: disable  Mi80OmFIVnBZMlhsdktEbHVwNDZhRFpPUlE9PTo1NzJlNDBmNw==

class ScenarioStepCreate(BaseModel):
    """创建步骤请求"""
    endpoint_id: Optional[UUID] = None
    name: str = Field(..., min_length=1, max_length=500)
    description: Optional[str] = None
    step_order: Optional[int] = None
    request_override: Dict[str, Any] = Field(default_factory=dict)
    headers_override: Dict[str, Any] = Field(default_factory=dict)
    extractors: List[StepExtractor] = Field(default_factory=list)
    assertions: List[StepAssertion] = Field(default_factory=list)
    condition_expression: Optional[str] = None
    continue_on_failure: bool = False
    delay_ms: int = Field(0, ge=0)
    retry_count: int = Field(0, ge=0)


class ScenarioStepUpdate(BaseModel):
    """更新步骤请求"""
    name: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = None
    request_override: Optional[Dict[str, Any]] = None
    headers_override: Optional[Dict[str, Any]] = None
    extractors: Optional[List[StepExtractor]] = None
    assertions: Optional[List[StepAssertion]] = None
    condition_expression: Optional[str] = None
    continue_on_failure: Optional[bool] = None
    delay_ms: Optional[int] = Field(None, ge=0)
    retry_count: Optional[int] = Field(None, ge=0)


class ScenarioStepResponse(ScenarioStepCreate):
    """步骤响应"""
    id: UUID
    scenario_id: UUID
    endpoint_id: Optional[UUID]
    step_order: int
    data_mappings: List[DataMappingResponse] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ==================== 执行模型 ====================

class ScenarioExecuteRequest(BaseModel):
    """执行场景请求"""
    variables: Dict[str, Any] = Field(default_factory=dict, description="运行时变量")
    base_url: str = Field("", description="API 基础 URL")
    async_mode: bool = Field(False, description="是否异步执行")


class ScenarioExecuteResponse(BaseModel):
    """执行场景响应"""
    run_id: UUID
    status: str
    message: str
    result: Optional["ScenarioRunResponse"] = None


class ScenarioRunResponse(BaseModel):
    """执行记录响应"""
    id: UUID
    scenario_id: UUID
    identifier: str
    status: str
    total_steps: int
    passed_steps: int
    failed_steps: int
    skipped_steps: int
    duration_ms: Optional[int]
    error_message: Optional[str]
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    created_at: datetime
# fmt: off  My80OmFIVnBZMlhsdktEbHVwNDZhRFpPUlE9PTo1NzJlNDBmNw==

    class Config:
        from_attributes = True


class ScenarioRunListResponse(BaseModel):
    """执行记录列表响应"""
    total: int
    items: List[ScenarioRunResponse]


class ScenarioStepResultResponse(BaseModel):
    """步骤执行结果响应"""
    id: UUID
    run_id: UUID
    step_id: UUID
    step_order: int
    status: str
    duration_ms: Optional[int]
    error_message: Optional[str]
    request_data: Optional[Dict[str, Any]]
    response_data: Optional[Dict[str, Any]]
    extracted_data: Dict[str, Any]
    assertion_results: List[Dict[str, Any]]

    class Config:
        from_attributes = True


# 更新前向引用
ScenarioExecuteResponse.model_rebuild()

File: ui/types/scenario.ts
/**
 * 版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。
 * 
 * 本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
 * 不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。
 * 
 * 授权商业应用请联系微信：huice666
 */

/**
 * 场景测试类型定义
 */
// FIXME  MC80OmFIVnBZMlhsdktEbHVwNDZkVXhWT1E9PTo3OGM2MzE1Yw==

export interface Scenario {
  id: string;
  project_id: string;
  identifier: string;
  name: string;
  description: string | null;
  folder_id: string | null;
  status: 'draft' | 'active' | 'archived';
  global_variables: Record<string, any>;
  retry_count: number;
  timeout_seconds: number;
  parallel_execution: boolean;
  total_steps: number;
  last_run_status: string | null;
  last_run_at: string | null;
  created_at: string;
  updated_at: string;
  created_by: string | null;
}
// @ts-expect-error  MS80OmFIVnBZMlhsdktEbHVwNDZkVXhWT1E9PTo3OGM2MzE1Yw==

export interface ScenarioStep {
  id: string;
  scenario_id: string;
  endpoint_id: string | null;
  step_order: number;
  name: string;
  description: string | null;
  request_override: Record<string, any>;
  headers_override: Record<string, any>;
  extractors: StepExtractor[];
  assertions: StepAssertion[];
  condition_expression: string | null;
  continue_on_failure: boolean;
  delay_ms: number;
  retry_count: number;
  data_mappings: DataMapping[];
  created_at: string;
  updated_at: string;
  endpoint?: {
    id: string;
    method: string;
    path: string;
    display_name: string;
  };
}

export interface StepExtractor {
  name: string;
  path: string;
  type?: string;
}

export interface StepAssertion {
  type: 'status' | 'jsonpath' | 'header';
  expected: any;
  path?: string;
  operator?: string;
}
// FIXME  Mi80OmFIVnBZMlhsdktEbHVwNDZkVXhWT1E9PTo3OGM2MzE1Yw==

export interface DataMapping {
  id: string;
  source_type: 'previous_response' | 'variable' | 'static';
  source_step_id: string | null;
  source_path: string | null;
  target_path: string;
  transform_expression: string | null;
  description: string | null;
}

export interface ScenarioRun {
  id: string;
  scenario_id: string;
  project_id: string;
  identifier: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
  runtime_variables: Record<string, any>;
  execution_config: Record<string, any>;
  total_steps: number;
  passed_steps: number;
  failed_steps: number;
  skipped_steps: number;
  duration_ms: number | null;
  report_path: string | null;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  executed_by: string | null;
}

export interface StepResult {
  id: string;
  run_id: string;
  step_id: string;
  endpoint_id: string | null;
  step_order: number;
  status: 'passed' | 'failed' | 'skipped' | 'error';
  request_data: Record<string, any> | null;
  response_data: Record<string, any> | null;
  extracted_data: Record<string, any>;
  assertion_results: AssertionResult[];
  duration_ms: number | null;
  error_message: string | null;
  error_stack: string | null;
  created_at: string;
}

export interface AssertionResult {
  assertion: StepAssertion;
  passed: boolean;
  actual: any;
  expected: any;
  message: string;
}
// @ts-expect-error  My80OmFIVnBZMlhsdktEbHVwNDZkVXhWT1E9PTo3OGM2MzE1Yw==
